packages/pyright-internal/src/languageService/completionProvider.ts

```diff
@@ -1427,6 +1456,299 @@ export class CompletionProvider {
                 }
             }
         });
+
+        if (baseExpressionNode === undefined) {
+            return;
+        }
+
+        if (isClassInstance(type) || isUnion(type)) {
+            const typedDicts = this._getTypedDicts(baseExpressionNode, parseNode, type);
+            if (typedDicts.length >= 1) {
+                // ignore already defined keys
+                const excludes = new Set(completionList.items.map((i) => i.label));
+                this._getDictExpressionStringKeys(parseNode).forEach((key) => {
+                    excludes.add(key);
+                });
+                typedDicts.forEach((typedDict) => {
+                    getTypedDictMembersForClass(this._evaluator, typedDict, /* allowNarrowed */ true).forEach(
+                        (_, key) => {
+                            // unions of TypedDicts may define the same key
+                            if (excludes.has(key)) {
+                                return;
+                            }
+                            excludes.add(key);
+
+                            this._addStringLiteralToCompletionList(
+                                key,
+                                quoteValue ? quoteValue.stringValue : undefined,
+                                postText,
+                                quoteValue
+                                    ? quoteValue.quoteCharacter
+                                    : this._parseResults.tokenizerOutput.predominantSingleQuoteCharacter,
+                                completionList
+                            );
+                        }
+                    );
+                });
+            }
+        }
+    }
+
+    private _getTypedDicts(baseExpressionNode: ExpressionNode, parseNode: ParseNode, type: ClassType | UnionType) {
+        if (!(tokenIsAtKey(this) || this._isInDictionaryStringKey(parseNode))) {
+            return [];
+        }
+
+        // path to resolve nested dictionaries
+        let path: DictPathEntry[] = [];
+        let curNode: ParseNode | undefined = parseNode;
+
+        while (curNode && curNode.id !== baseExpressionNode.id) {
```

Comment:
rather than resolving containing types using parse tree, can you just bind expression of where the key is used and find one what type you are supposed to use?

Reply:
I guess that didn't work? and that is why you have to find parameter type and dig in to find type of outter key type?

Reply:
this seems good helper method to put in parse tree utils as well if just binding containing expression can't figure out what outer typeddict current expression is in. and we have to do this search up to find call and find parameter type and dig in to find type of key of nested typed dict.

Reply:
Yeah, a language service provider ideally shouldn't need to do so much parse tree walking to try to determine type information. That's the job of the type evaluator. Currently, the type evaluator doesn't expose a way to extract an "expected type" for a given node, but this is something that I could look at adding. This would allow us to remove all of this complex code from the completion provider. It would also allow us to support much more complex scenarios where expressions are deeply nested (e.g. a TypedDict within a TypedDict within a list within a tuple). Give me a few hours to investigate the feasibility of exposing this sort of functionality from the type evaluator.

Reply:
@RobertCraigie, here's a PR that adds a new entry point to the typeEvaluator called getExpectedType. It allows you to ask what the expected type is for a specified node. I think this would allow you to remove a bunch of code from this PR and make it work in the more general case (including nested expressions). https://github.com/microsoft/pyright/pull/2336

Reply:
@erictraut Thank you! I've given it a shot and it works perfectly for non-nested TypedDicts but I can't figure out which node I should call `getExpectedType` on to resolve nested expressions.

For example, given `thing3({'wrapped': {'[|/*marker13*/|]'}})` and calling `getExpectedType` on the `Set` node representing `{'[|/*marker13*/|]'}` resolves to `Any`. I also tried the node at the marker and all of its parents but they also resolved to `Any`. Am I doing something wrong?

Reply:
Think I've figured out why the issue mentioned above is happening, it's to do with how `evaluateTypesForExpressionInContext` works with call expressions as explicitly assigning a variable to a TypedDict correctly resolves the nested TypedDict type.

```py
user: User = {
    'meta': {
        '|'
    }
}
```

Reply:
I can take a look at it if you commit and push your updated code that uses `getExpectedType`. I would expect the case covered by `marker13` to work, so if it's not working, I'd like to debug it.

Reply:
@erictraut thank you, I've pushed my WIP code that makes use of `getExpectedType`.